File: solutions/145/c++/solution.cpp
// 145. Binary Tree Postorder Traversal
#include <vector>
#include <stack>

using namespace std;

class Solution
{
	public:
	vector<int> postorderTraversal(TreeNode* root)
	{
		vector<int> postorderValues;

		if (root == NULL)
		{
			return postorderValues;
		}

		stack<TreeNode*> nodes;
		TreeNode* previousNode = NULL;

		nodes.push(root);

		while (!nodes.empty())
		{
			TreeNode* currentNode = nodes.top();

			if (previousNode != NULL && (currentNode->left == previousNode || currentNode->right == previousNode) || previousNode == currentNode)
			{
				postorderValues.push_back(currentNode->val);
				nodes.pop();
			}
			else
			{
				if (currentNode->right != NULL)
				{
					nodes.push(currentNode->right);
				}

				if (currentNode->left != NULL)
				{
					nodes.push(currentNode->left);
				}
			}

			previousNode = currentNode;
		}

		return postorderValues;
	}
};

File: solutions/145/java/Solution.java
// 145. Binary Tree Postorder Traversal
import java.util.Deque;
import java.util.LinkedList;
import java.util.List;

class Solution
{
	public List<Integer> postorderTraversal(TreeNode root)
	{
		List<Integer> postorder = new LinkedList<>();
		Deque<TreeNode> nodesLeft = new LinkedList<>();

		nodesLeft.addFirst(root);

		while (!nodesLeft.isEmpty())
		{
			TreeNode currentNode = nodesLeft.removeFirst();

			if (currentNode == null)
			{
				continue;
			}

			postorder.add(0, currentNode.val);

			nodesLeft.addFirst(currentNode.left);
			nodesLeft.addFirst(currentNode.right);
		}

		return postorder;
	}
}
